services/meta/handler.go

```diff
@@ -0,0 +1,238 @@
+package meta
+
+import (
+	"compress/gzip"
+	"expvar"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/influxdb/influxdb/uuid"
+)
+
+// Handler represents an HTTP handler for the InfluxDB server.
+type Handler struct {
+	config  *Config
+	Version string
+
+	Logger         *log.Logger
+	loggingEnabled bool // Log every HTTP access.
+	pprofEnabled   bool
+	store          *store
+}
+
+// NewHandler returns a new instance of handler with routes.
+func NewHandler(c *Config, s *store) *Handler {
+	h := &Handler{
+		config:         c,
+		Logger:         log.New(os.Stderr, "[meta-http] ", log.LstdFlags),
+		loggingEnabled: c.LoggingEnabled,
+		store:          s,
+	}
+
+	return h
+}
+
+// SetRoutes sets the provided routes on the handler.
+func (h *Handler) WrapHandler(name string, hf http.HandlerFunc) http.Handler {
+	var handler http.Handler
+	handler = http.HandlerFunc(hf)
+	handler = gzipFilter(handler)
+	handler = versionHeader(handler, h)
+	handler = cors(handler)
+	handler = requestID(handler)
+	if h.loggingEnabled {
+		handler = logging(handler, name, h.Logger)
+	}
+	handler = recovery(handler, name, h.Logger) // make sure recovery is always last
+
+	return handler
+}
+
+// ServeHTTP responds to HTTP request to the handler.
+func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	switch r.Method {
+	case "HEAD":
+		h.WrapHandler("ping", h.servePing).ServeHTTP(w, r)
+	case "GET":
+		h.WrapHandler("cache", h.serveCache).ServeHTTP(w, r)
+	case "POST":
+		h.WrapHandler("execute", h.serveExec).ServeHTTP(w, r)
+	default:
+		http.Error(w, "bad request", http.StatusBadRequest)
+	}
+}
+
+// serveExec executes the requested command.
+func (h *Handler) serveExec(w http.ResponseWriter, r *http.Request) {
+}
+
+// serveCache is a long polling http connection to server cache updates
+func (h *Handler) serveCache(w http.ResponseWriter, r *http.Request) {
+	// get the current index that client has
+	clientIndex, _ := strconv.Atoi(r.URL.Query().Get("index"))
+	index, c := h.store.Snapshot()
+
+	// If the client has an older index, send the updated cache
+	if index > clientIndex {
+		w.Header().Add("influxdb-meta-index", strconv.Itoa(index))
+		w.Write(c)
+		return
+	}
+
+	// Make sure if the client disconnects we signal the query to abort
+	closing := make(chan struct{})
+	if notifier, ok := w.(http.CloseNotifier); ok {
+		notify := notifier.CloseNotify()
+		go func() {
+			<-notify
+			close(closing)
+		}()
+	}
+
+	// block until we get a data change, or an error which signals
+	// scenarios like leader change, store closing, etc.
+	h.store.WaitForDataChanged(closing)
+
+	index, c = h.store.Snapshot()
+	w.Header().Add("influxdb-meta-index", strconv.Itoa(index))
+	w.Write(c)
+	return
+}
```

Comment:
Ah, yes, agree with all of that.  This was coded this way to make it easier to get the long polling mocked out.  It's also easy enough to add those checks for content-type as well.